### Diabetes prediction model using ANN

In [128]:
import numpy as np
import pandas as pd

from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, accuracy_score


import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers #layers will help us in developing the neural network layers

In [129]:
df = pd.read_csv('/diabetes.csv')

In [130]:
df.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


In [131]:
df.shape

(768, 9)

In [132]:
X = df.drop('Outcome', axis=1)
y = df['Outcome']

In [133]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
dtypes: float64(2), int64(6)
memory usage: 48.1 KB


In [134]:
#Check null/ missing data
X.isnull().sum()

,0
Pregnancies,0
Glucose,0
BloodPressure,0
SkinThickness,0
Insulin,0
BMI,0
DiabetesPedigreeFunction,0
Age,0


In [135]:
X.isna().sum()

,0
Pregnancies,0
Glucose,0
BloodPressure,0
SkinThickness,0
Insulin,0
BMI,0
DiabetesPedigreeFunction,0
Age,0


In [136]:
X.describe()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000


In [137]:
#Split data into trainand test
X_train, X_test,y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [138]:
X_train.shape, X_test.shape

((614, 8), (154, 8))

In [139]:
X_train.head(5)

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age
60,2,84,0,0,0,0.0,0.304,21
618,9,112,82,24,0,28.2,1.282,50
346,1,139,46,19,83,28.7,0.654,22
294,0,161,50,0,0,21.9,0.254,65
231,6,134,80,37,370,46.2,0.238,46


In [140]:
#Scale numeric columns
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [141]:
#Build ANN model
model = keras.Sequential([
    layers.Input(shape=(X_train.shape[1],),name='Input_layer'), # Input layer has 8 neurons
    layers.Dense(16,activation='relu',name='Hidden_1'), # Hidden Layer 1 with 16 neurons
    layers.Dense(32,activation='relu',name='Hidden_2'), # Hidden Layer 2 with 32 neurons
    layers.Dense(64,activation='relu',name='Hidden_3'), # Hidden Layer 3 with 64 neurons
    layers.Dense(1,activation='sigmoid',name='Output') #Output Layer : The Sigmoid activation function converts the output into a probability between 0 and 1
])

In [142]:
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

In [143]:
model.fit(X_train,y_train,epochs=15,batch_size=32)

Epoch 1/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.6808 - loss: 0.6486
Epoch 2/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.7068 - loss: 0.5808
Epoch 3/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.7394 - loss: 0.5330
Epoch 4/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.7606 - loss: 0.5021
Epoch 5/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.7671 - loss: 0.4779
Epoch 6/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.7736 - loss: 0.4646
Epoch 7/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.7834 - loss: 0.4540
Epoch 8/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.7866 - loss: 0.4463
Epoch 9/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7850 - loss: 0.4411
Epoch 10/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7964 - loss: 0.4345
Epoch 11/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7948 - loss: 0.4312
Epoch 12/15
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8046

In [144]:
y_pred_proba =model.predict(X_test)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step


In [145]:
# Convert probabilities to 0/1 labels
y_pred = (y_pred_proba > 0.5).astype(int)

In [127]:
#Plot Confusion Matrics and the accuracy of the model
cm = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(cm)

tn, fp, fn, tp = cm.ravel()

print(f"TN: {tn}")
print(f"FP: {fp}")
print(f"FN: {fn}")
print(f"TP: {tp}")

print("Accuracy:", accuracy_score(y_test, y_pred))

Confusion Matrix:
[[81 18]
 [20 35]]
TN: 81
FP: 18
FN: 20
TP: 35
Accuracy: 0.7532467532467533


##### The ANN Model achieved 75% accuracy on test data
1. The model correctly predicts 116 members out of 154 members for diabetes diagnosis.
2. The model predicts 18 false negatives i.e. 18 healthy members are predicted as diabetic.
3. 20 diabetic patients are falsely predicted as healthy members. In Healtcare models, it is important to lower the false negative prediction to avoid the delay in diagnosis and treatment.